witness/golang/cmd/witness/internal/witness/witness.go

```diff
@@ -219,6 +244,52 @@ func (w *Witness) getLatestChkpt(queryRow func(query string, args ...interface{}
 	return maxChkpt, nil
 }
 
+// getLatestChkptAndRange returns the latest checkpoint and corresponding
+// compact range for a given log.
+func (w *Witness) getLatestChkptAndRange(queryRow func(query string, args ...interface{}) *sql.Row, logID string) (Chkpt, [][]byte, error) {
+	row := queryRow("SELECT raw, size, range FROM chkpts WHERE logID = ?", logID)
+	if err := row.Err(); err != nil {
+		return Chkpt{}, nil, err
+	}
+	var maxChkpt Chkpt
+	var rawPf []byte
+	if err := row.Scan(&maxChkpt.Raw, &maxChkpt.Size, &rawPf); err != nil {
+		if err == sql.ErrNoRows {
+			return Chkpt{}, nil, status.Errorf(codes.NotFound, "no checkpoint for log %q", logID)
+		}
+		return Chkpt{}, nil, err
+	}
+	var pf log.Proof
+	if err := pf.Unmarshal(rawPf); err != nil {
+		return Chkpt{}, nil, fmt.Errorf("couldn't unmarshal proof: %v", err)
+	}
+	return maxChkpt, pf, nil
+}
+
+// verifyRange verifies the new checkpoint against the stored and given compact
+// range and outputs the new compact range if verification succeeds.
+func (w *Witness) verifyRange(chkpt *log.Checkpoint, oldSize uint64, h hashers.LogHasher, curRangeRaw [][]byte, newRangeRaw [][]byte) ([][]byte, error) {
+	rf := compact.RangeFactory{Hash: h.HashChildren}
+	curRange, err := rf.NewRange(0, oldSize, curRangeRaw)
+	if err != nil {
+		return nil, fmt.Errorf("can't form current compact range: %v", err)
+	}
```

Comment:
Worth doing a `GetRootHash` on curRange and double checking it's valid for the old chkpt? (would help debugging in the case that updates were failing because the range in the DB was corrupt rather than the incoming proof being wrong, etc.)

Reply:
Yup, good idea!  It's also currently possible that for the first checkpoint the caller gives complete garbage for the range, which we wouldn't find out until we went to use it.  So I guess it would be good to also add a validity check before storing it (this is line 178 of `witness.go`, where currently I just store it without any checks).

Reply:
👍

Reply:
Okay, that's done now!

Reply:
This is done now!